### DISCUSSION TITLE
userId of player changes

### DISCUSSION THREAD

**Discussion by RaphaelHippe** [saul/demofile#221] [Questions] [answered]
Hello, 

first of all I'd like to thank you for this awesome library. 

I am currently working with this GOTV demofile: https://www.hltv.org/matches/2317947/mousesports-vs-faze-ecs-season-4-finals

and noticed that the userId for rain seems to change right at the end of the game: 

for every kill / assist / death for rain it is: 187

when the 'end' event is fired I am checking for all userIds: 
```javascript
demoFile.on('end', () => {
      let testPlayerIds = demoFile.players.map(p => p.userId);
      console.log('testPlayerIds', testPlayerIds);
```

and get the following output: 
```
testPlayerIds [ 176, 177, 178, 179, 192, 181, 182, 183, 189, 185, 190, 188, 191 ]
```
and the 187 is missing. 

I left in the example code for notification when a player joins: 

```javascript
 demoFile.entities.on('create', e => {
      if (e.entity.serverClass.name !== 'CCSPlayer') {
        return;
      }
      if (e.entity.userInfo) {
        console.log('%s (%s) joined the game', e.entity.name, e.entity.steamId);
      }
    });
```

and don't get any console prints mid game for rain rejoining the game. Therefore it seems like the userId is changing mid game for no obvious reason. 

Do you have any idea why that would be the case? 

Thanks, 
Raphael Hippe

**Comment by saul** [comment]
UserIDs certainly don't change throughout the game - are you sure rain didn't disconnect before the game ended?

**Comment by RaphaelHippe** [comment]
Hey, 

I just now tested further for disconnects: 

```javascript
    demoFile.entities.on('beforeremove', e => {
      if (e.entity.userInfo) {
        console.log('user entity gets removed:');
        console.log(e.entity.name, e.entity);
      }
    });
```

```javascript
    demoFile.gameEvents.on('client_disconnect', e => {
      console.log('a client disconnected');
      let testPlayerIds = demoFile.players.map(p => p.userId);
      console.log('testPlayerIds', testPlayerIds);
    });
```

while the event `client_disconnect` never fires, the `beforeremove` shows me that rains entity indeed gets deleted right before the `end` event of the demoFile gets fired. 

So I am not quite sure how to precede from here. 

my goal is to get all actual player Ids (onle the 10 who play and not any players, that spectate or something which `demoFile.players.map(p => p.userId);` seems to include) 

I tried to look at the `e.entity` for the `beforeremove` event but I didn't find an ID in there. And since the entity just gets remove I couldn't override my current ID for rain anyhow.

**Comment by saul** [comment]
You can check if a player is on a team by checking their team number is 2 or 3. 0 = Unassigned, 1 = Spectator, 2 = Terrorist and 3 = CT.

Also `player_disconnect` is the event you're looking for.

If you want to take stats at the end of the game, you can check when the game phase switches to match end:

```js
  demo.entities.on('DT_CSGameRules.m_gamePhase', e => {
    if (e.newValue == 5) {
      // Game finished
    }
  });
```

**Comment by RaphaelHippe** [comment]
Thank you!

for game events I am using this reference: https://wiki.alliedmods.net/Counter-Strike:_Global_Offensive_Events

and `player_disconnect` is not on that reference. Do you have one thats more up to date?

Cheers

**Comment by saul** [comment]
There is also https://wiki.alliedmods.net/Generic_Source_Events and https://wiki.alliedmods.net/Generic_Source_Server_Events - see the link at the top of the CSGO events page.

Thanks

________________________________

**Comment by RaphaelHippe** [comment]
Hey, 
sadly, `DT_CSGameRules.m_gamePhase` doesn't get fired at all when I am listening on `demoFile.entities`

code block: 

```javascript
demoFile.entities..on('DT_CSGameRules.m_gamePhase', e => {
  console.log('TEST', e);
  if(e.newValue == 5) {
    console.log('TEST 2');
  }
});
```

Cheers

**Comment by saul** [comment, marked as answer]
Sorry, that should have been ‘change’ and you should be checking `e.tableName` and `e.varName`

**Comment by RaphaelHippe** [comment]
Thanks, that works :-)
